staging/src/k8s.io/client-go/util/namespace/namespace.go

```diff
@@ -0,0 +1,41 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"io/ioutil"
+	"os"
+	"strings"
+)
+
+// Namespace returns the namespace in inCluster context
+func Namespace() string {
```

Comment:
I don't think this makes sense standalone, especially with the fallback to "default", with no indication whether that was positively found or whether it is a fallback. In the context of inClusterClientConfig{}, there's a `Possible()` method that tells you if you are even in an in-cluster context, and the icc namespace isn't used if it is not in that context.

Reply:
something like this in client_config.go might be better:

```
func NewInClusterClientConfig() InClusterConfig {
	return &inClusterClientConfig{inClusterConfigProvider: restclient.InClusterConfig}
}
```

used like this:
```
icc := clientcmd.NewInClusterClientConfig()
config, err := icc.ClientConfig()
if err != nil {
  ... handle error from not actually being in-cluster ...
  return
}
... construct client from config ...
currentNamespace := icc.Namespace()
```

Reply:
The idea is to be able to use this logic without importing Kubernetes, to make a controller by example.
Not fallback to `default` is fine too, we don't care, we only want it in an "inCluster" context, making `inClusterClientConfig` public would work too.

Reply:
I'm quite against having this code only in `InClusterConfig` has we often use `clientcmd.BuildConfigFromFlags` which return a `restclient.Config`.

I prefer to handle a `""`/false return in a single function than to do abstraction around in/out cluster only to manage a single namespace string.